dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,260 @@
+
+    @VisibleForTesting
+    PinUntilErrorChannel(NodeList nodeList) {
+        this.maybeShufflingNodeList = nodeList;
+    }
+
+    public static PinUntilErrorChannel pinUntilError(List<LimitedChannel> channels) {
+        Preconditions.checkArgument(!channels.isEmpty(), "List of channels must not be empty");
+        ReshufflingNodeList shufflingNodeList =
+                new ReshufflingNodeList(channels, ThreadLocalRandom.current(), System::nanoTime);
+        return new PinUntilErrorChannel(shufflingNodeList);
+    }
+
+    /**
+     * This strategy results in unbalanced server utilization, especially after an upstream restarts.
+     * @deprecated prefer {@link #pinUntilError}
+     */
+    @Deprecated
+    public static PinUntilErrorChannel pinUntilErrorWithoutReshuffle(List<LimitedChannel> channels) {
+        Preconditions.checkArgument(!channels.isEmpty(), "List of channels must not be empty");
+        ConstantNodeList constantNodeList = new ConstantNodeList(channels, ThreadLocalRandom.current());
+        return new PinUntilErrorChannel(constantNodeList);
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        int currentIndex = currentHost.get();
+        LimitedChannel channel = maybeShufflingNodeList.get(currentIndex);
+
+        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        if (!maybeFuture.isPresent()) {
+            OptionalInt next = incrementCurrentHost(currentIndex);
+            debugLogCurrentChannelRejected(currentIndex, channel, next);
+            return Optional.empty(); // if the caller retries immediately, we'll get the next host
+        }
+
+        ListenableFuture<Response> future = maybeFuture.get();
+        return Optional.of(DialogueFutures.addDirectCallback(future, new FutureCallback<Response>() {
+            @Override
+            public void onSuccess(Response response) {
+                if (response.code() >= 300) {
+                    OptionalInt next = incrementCurrentHost(currentIndex);
+                    debugLogReceivedErrorStatus(currentIndex, channel, response, next);
+                    // TODO(dfox): handle 308 See Other somehow, as we currently don't have a host -> channel mapping
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                OptionalInt next = incrementCurrentHost(currentIndex);
+                debugLogReceivedThrowable(currentIndex, channel, throwable, next);
+            }
+        }));
+    }
+
+    /**
+     * If we have some reason to think the currentIndex is bad, we want to move to the next host. This is done with a
+     * compareAndSet to ensure that out of order responses which signal information about a previous host don't kick
+     * us off a good one.
+     */
+    private OptionalInt incrementCurrentHost(int currentIndex) {
```

Comment:
maybeIncrementAndGet ? kinda wish you had parens on that i.e. maybe(IncrementAndGet)

Reply:
What about if I just called it `markCurrentHostAsBad` ?

Reply:
It doesn't quite capture the return type though. `maybeIncrementCurrentHost`?